.pre-commit-config.yaml

```diff
@@ -0,0 +1,59 @@
+# Pre-commit is tool which allows to run checks automatically before commit.
+# This way commits are always consistent and clean.
+
+exclude: target/|\.git/|.*\.snap
+
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: c4a0b883114b00d8d76b479c820ce7950211c99b # v4.5.0
+    hooks:
+      - id: check-merge-conflict
+        args: [--assume-in-merge]
+      # Editorconfig-checker does not care if file contains multiple newlines.
+      - id: end-of-file-fixer
+      - id: check-yaml
+      - id: check-case-conflict
+      - id: trailing-whitespace
+      - id: fix-byte-order-marker
+
+  - repo: https://github.com/Lucas-C/pre-commit-hooks
+    rev: 762c66ea96843b54b936fc680162ea67f85ec2d7 # v1.5.4
+    hooks:
+      - id: remove-tabs
+        name: Remove tabs (4 spaces)
+        args: ['--whitespaces-count', '4']
+        exclude: \.(yaml|yml)$
+      - id: remove-tabs
+        name: Remove tabs (2 spaces)
+        args: ['--whitespaces-count', '2']
+        files: \.(yaml|yml)$
+
+  - repo: https://github.com/pre-commit/pygrep-hooks
+    rev: 3a6eb0fadf60b3cccfd80bad9dbb6fae7e47b316 # v1.10.0
+    hooks:
+      - id: text-unicode-replacement-char
+        exclude: ".*test_fdx_utf8.cpp$"
+
+  # Formatting Python code. We have not so many files so no example flake8 yet.
+  - repo: https://github.com/psf/black
+    rev: 2a1c67e0b2f81df602ec1f6e7aeb030b9709dc7c #23.11.0
+    hooks:
+      - id: black
+        language_version: python3
+
+  - repo: https://github.com/doublify/pre-commit-rust
```

Comment:
I find this good for PRs but we often commit a wip state to a branch where this would always end up being denied.
This is especially true for refactors where we might not yet be in a working state in the commit.

Reply:
Yeah and that is totally fine. That is why we can do

`git commit --no-verify`

I normally first still try to do git commit and it will clean some things and if I really need I use --no-verify. I can of course add some documentation about these things. Maybe create CONTRIBUTING file as we do not have that eather.

Reply:
> Yeah and that is totally fine. That is why we can do
> 
> `git commit --no-verify`
> 
> I normally first still try to do git commit and it will clean some things and if I really need I use --no-verify. I can of course add some documentation about these things. Maybe create CONTRIBUTING file as we do not have that eather.
> 
> Currently pre-commit does not prevent pushing. Check are only made in commit phase.

I think CONTRIBUTING is a good idea i'll look into that and a Code of Conduct. We're getting much more interest recently than when we first started :)